src/bio/bio_buffer.c

```diff
@@ -637,8 +637,18 @@ rw_completion(void *cb_arg, int err)
 	D_ASSERT(xs_ctxt->bxc_blob_rw > 0);
 	xs_ctxt->bxc_blob_rw--;
 
-	if (biod->bd_result == 0 && err != 0) {
-		biod->bd_result = daos_errno2der(-err);
+	/* Induce NVMe Read/Write Error*/
+	if (biod->bd_update)
+		err = DAOS_FAIL_CHECK(DAOS_NVME_WRITE_ERR) ? -DER_IO : err;
+	else
+		err = DAOS_FAIL_CHECK(DAOS_NVME_READ_ERR) ? -DER_IO : err;
+
+	/* Return the error value of the first NVMe IO error */
+	if (biod->bd_result == 0 && err != 0)
+		biod->bd_result = err;
```

Comment:
biod->bd_result = daos_errno2der(-err);

BTW: I found that daos_errno2der() and daos_der2errono() are mismatching, the default error code for daos_errno2der() is -DER_MISC, but the default error code of daos_der2errono() is EIO, also, EIO is converted to -DER_MISC in daos_errno2der(), I'm not sure if it's intentional. Sydney, any comments? I see this function was last changed by you.

Probably you'd fix it in this patch as well?

Reply:
@NiuYawei @ravalsam I checked git blame (on daos/common.h), seems like I made the updated on daos_errno2der in 2018 (699da51e) and the updates to daos_der2errno were made by Ashley in May 2019 (78b09bb8). There definitely is a mismatch here, I don't remember the reasoning for converting EIO to MISC here (maybe there wasn't DER_IO available 2 years ago?). I think the daos_errno2der is just outdated and needs to be updated.

Reply:
update the case for EIO in daos_errno2der()